### ISSUE TITLE
EffectiveViewportChanged and InvalidateMeasure/InvalidateArrange

### ISSUE THREAD

**Issue by xLEGiON** [AvaloniaUI/Avalonia#13182]
## Describe the bug

I initially discovered this when trying to have both a Avalonia.Labs.Lottie and Avalonia.Controls.TreeDataGrid on the same view. Sometimes it would work, but most of the time the Lottie animation wouldn't show and the TreeDataGrid would break (unable to scroll, etc). I managed to distill the problem down into two example controls that cause the same behavior.

## To Reproduce

```csharp
using System;
using System.Diagnostics;
using Avalonia.Controls;
using Avalonia.Interactivity;
using Avalonia.Layout;

namespace Controls;


public class ExampleControl : Control
{

    protected override void OnLoaded(RoutedEventArgs e)
    {
        Debug.WriteLine("OnLoaded");
        base.OnLoaded(e);
        LayoutUpdated += OnLayoutUpdated;
        InvalidateMeasure();
    }

    private void OnLayoutUpdated(object? sender, EventArgs e)
    {
        Debug.WriteLine("OnLayoutUpdated");
    }

}

public class ExampleControlTwo : Control
{

    public ExampleControlTwo()
    {
        EffectiveViewportChanged += ExampleControlTwo_EffectiveViewportChanged;
    }

    private void ExampleControlTwo_EffectiveViewportChanged(object? sender, EffectiveViewportChangedEventArgs e)
    {
        Debug.WriteLine("ExampleControlTwo_EffectiveViewportChanged");
    }

}
```
Put only ExampleControl in a view and its OnLayoutUpdated method will execute as expected. Add ExampleControlTwo to the same view and the OnLayoutUpdated method in ExampleControl will no longer execute.

## Expected behavior

The  OnLayoutUpdated method of ExampleControl should execute regardless of whether ExampleControlTwo is present in the same view.

## Environment

- OS: Windows 10
- Avalonia-Version: 11.0.5-rc1

## Additional context

I'm not yet familiar enough with the Avalonia source code to know what is causing this. I can see that adding to EffectiveViewportChanged adds a listener to the LayoutManager. This then affects the RaiseEffectiveViewportChanged method in the LayoutManager.

**Comment by MrJul** [comment]
Can you please provide a minimal reproducible example project with an attached view?

Because I can't reproduce it with the two controls you mentioned. Test window XAML:
```xaml
<Grid ColumnDefinitions="*,Auto,*">
  <controls:ExampleControl Grid.Column="0" />
  <GridSplitter Grid.Column="1" Background="Magenta" />
  <controls:ExampleControlTwo Grid.Column="2" />
</Grid>
```

Move the slider: the event is correctly raised and printed in the debug console.
`LayoutUpdated` should always be raised whenever a layout occurs: it's not conditional.

My first guess when reading your issue is that an uncaught exception happens somewhere with the controls you're using, crashing the layout.

**Comment by xLEGiON** [comment]
It looks like you're right. I should have tested with a fresh project. It only seems to be occurring on views that are navigated to with a FluentAvalonia Frame specifically using DrillInNavigationTransitionInfo. I'll close this for now and investigate further, raising an issue with FluentAvalonia if necessary. Thanks for your help.

**Comment by xLEGiON** [comment]
After digging deeper, this seems more like an Avalonia specific issue. The problem occurs when an animation is applied to the view containing both controls where ScaleTransform.ScaleXProperty or ScaleTransform.ScaleYProperty are set to 0 for the first keyframe.

MainWindow.axaml:
```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="AvaloniaApplication1.Views.MainWindow"
        Icon="/Assets/avalonia-logo.ico"
        Title="AvaloniaApplication1">
    <StackPanel>
        <ContentPresenter x:Name="TestContentPresenter" />
        <Button x:Name="TestButton">Test</Button>
    </StackPanel>
</Window>
```

MainWindow.axaml.cs:
```csharp
using Avalonia;
using Avalonia.Controls;
using Avalonia.Animation;
using Avalonia.Animation.Easings;
using Avalonia.Media;
using Avalonia.Styling;
using System;
using System.Threading;
using Avalonia.Threading;

namespace AvaloniaApplication1.Views
{
    public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();
            TestButton.Click += TestButton_Click;
        }

        private void TestButton_Click(object? sender, Avalonia.Interactivity.RoutedEventArgs e)
        {
            TestContentPresenter.Content = new MainView();
            CancellationTokenSource cts = new();
            Dispatcher.UIThread.Post(async () =>
            {
                var animation = new Animation
                {
                    Easing = new SplineEasing(0.1, 0.9, 0.2, 1.0),
                    Children =
                        {
                            new KeyFrame
                            {
                                Setters =
                                {
                                    new Setter(Visual.OpacityProperty, 0.0),
                                    new Setter(ScaleTransform.ScaleXProperty, 0.0),
                                    new Setter(ScaleTransform.ScaleYProperty, 0.0)
                                },
                                Cue = new Cue(0d)
                            },
                            new KeyFrame
                            {
                                Setters =
                                {
                                    new Setter(Visual.OpacityProperty, 1.0),
                                    new Setter(ScaleTransform.ScaleXProperty, 1.0),
                                    new Setter(ScaleTransform.ScaleYProperty, 1.0)
                                },
                                Cue = new Cue(1d)
                            }
                        },
                    Duration = TimeSpan.FromSeconds(0.67),
                    FillMode = FillMode.Forward
                };
                await animation.RunAsync(TestContentPresenter, cts.Token);
            }, DispatcherPriority.Render);
        }
    }
}
```

MainView.axaml:
```xaml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:c="using:Controls"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="AvaloniaApplication1.Views.MainView"
             Foreground="White"
             Background="#171C2C">
    <StackPanel>
        <TextBlock Text="Welcome" HorizontalAlignment="Center" VerticalAlignment="Center"/>
        <c:ExampleControl Width="50" Height="50" />
        <c:ExampleControlTwo Width="50" Height="50" />
    </StackPanel>
</UserControl>
```

LayoutUpdated for ExampleControl isn't raised with the above code. However, if you change either ScaleXProperty or ScaleYProperty in the first keyframe to 0.01, it works.

Possibly related to https://github.com/AvaloniaUI/Avalonia/issues/12452

**Comment by MrJul** [comment]
Thanks for providing a reproduction!

Indeed, this is exactly #12452, I can see a first chance `System.InvalidOperationException`: *Transform is not invertible* being thrown. (As also mentioned in that same issue, the exception is lost somewhere instead of being rethrown.)

**Comment by MrJul** [comment]
Closing as a duplicate of #12452, which should be fixed by #13188.
